File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n= (int)nums.size();
        set<vector<int>> vis;
        unordered_set<int> ump;
        vector<int> temp;
        for(int i=0;i<=n-2;++i){
            for(int j=i+1;j<n;++j){
                if(ump.find(-nums[i]-nums[j])!=ump.end()){
                    temp={nums[i],nums[j],-nums[i]-nums[j]};
                    sort(temp.begin(),temp.end());
                    vis.insert(temp);
                }
            }
            if(ump.find(nums[i])==ump.end()) ump.insert(nums[i]);
        }
        return vector<vector<int>> ({vis.begin(),vis.end()});
    }
};

File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Set<List<Integer>> store = new HashSet<>();
        Arrays.sort(nums);
        for(int i=0;i<n-2;i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            int tar= -nums[i];
            int left = i+1, right = n-1;
            while(left<right){
                int sum = nums[left]+nums[right];
                if(sum<tar){
                    ++left;
                }
                else if(sum>tar){
                    --right;
                }
                else{
                    store.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    ++left;
                    --right;
                }
            }
        }
        return new ArrayList<>(store);
    }
}
